### PR TITLE
Schema: add MAPEM and SPATEM JSON schemas + fix DSRC references for SREM and SSEM

### LINKED ISSUE #150: Define SPATEM and MAPEM message contracts
GIVEN I decided to create a new ITS message among SPATEM or MAPEM
WHEN I use the provided JSON schema file as specification to build the message
THEN I will be able to validate the format

- [ ] SPATEM
- [ ] MAPEM
- [ ] align the version for the first packaging
- [ ] validate the contracts into the GitHub workflow
- [ ] Specify the documents used as reference to create the schemas

### LINKED ISSUE #234: schemas: DSRC schema is in schemas' top-level directory
**What:**

* the DSRC schema is at the top-level of schemas' directory, `schema/dsrc_schema_1-0-0.json`;
* the references in schemas are now wrong: the DSRC is not in the same directory as the schemas.

---
**Expected:**

* the DSRC schema is in a sub-directory of its own, i.e. `schema/dsrc/dsrc_schema_1-0-0.json`;
* the references in schemas point to `../dsrc/dsrc_schema_1-0-0.json`.

### PR DESCRIPTION
**Pull request by mathieu1fb**
**What's fixed**;

JSON Schemas:
- Moved DSRC schema from top-level to a dedicated directory
- Fixed SREM and SSEM references to DSRC definitions

closes #234

---
**What's new**;

JSON Schemas:
- Added MAPEM and SPATEM schemas.
- Updated the DSRC schema in a new 2.0.0 version, with additional fields needed for these two messages.

closes #150 

---
**What to do**:

Read the new schemas and check them against [ETSI repository docs](https://forge.etsi.org/rep/ITS/asn1/is_ts103301/-/tree/release2/docs) for potential mistakes.

### CODE REVIEW

**Comment by mathieu1fb** [comment]
Should I close issue #234 with this PR, or make a dedicated one later on?

**Comment by Hugues360** [comment]
> Should I close issue #234 with this PR, or make a dedicated one later on?

If @ymorin-orange agrees, I think it will be easier to close issue #234 in this PR in order that this PR fixes 2 issues

**Comment by ymorin-orange** [comment]
> Should I close issue #234 with this PR, or make a dedicated one later on?

Ideally, #234 should be fixed first in a separate branch, then this one rebased afterwards. **Ideally.**

However, I think it would be OK if #234 was to be fixed in this MR. But then, the fix **shall be the first commit in the series**.

**Comment by mathieu1fb** [comment]
I've added the new `$references` section to the 2.0.0 versions of MAPEM, SPATEM and DSRC JSON schemas.

**Comment by mathieu1fb** [comment]
A remark / interrogation on my side:

Is it better to keep bit strings such as:
```
"directional_use": {
  "type": "string",
  "description": "2-bit string. Bits used: ingressPath (0), egressPath (1). Denote the allowed direction of travel over a lane object. By convention, the lane object is always described from the stop line outwards away from the intersection. Therefore, the ingress direction is from the end of the path to the stop line and the egress direction is from the stop line outwards.",
  "examples": [
    "01",
    "10",
    "11"
  ]
}
```
Or should we go instead for something more verbose like:
```
"directional_use": {
  "type": "object",
  "description": "Denote the allowed direction of travel over a lane object. By convention, the lane object is always described from the stop line outwards away from the intersection.",
  "properties": {
    "ingress_path": {
      "type": "boolean",
      "description": "The ingress direction is from the end of the path to the stop line."
    },
    "egress_path": {
      "type": "boolean",
      "description": "The egress direction is from the stop line outwards."
    }
  }
},
```
Knowing that some fields have a dozen different sub-fields.

**Comment by ymorin-orange** [comment]
> Is it better to keep bit strings such as:
> [bit-wise description]
> Or should we go instead for something more verbose like:
> [per-bit boolean property]
> Another alternative would be a list of enum:
> [enum]

It was my understanding that we would stick the closest possible to the ASN.1 spec, so as to still claim we are doing ETSI messages. But in such a case I can see how cymbersome it is...

The solution with a boolean property per bit is the best of the two; it is explicit and fits the spirit of a schema.

The enum proposal does looks kludgy and half-backed; it looks like the ETSI string-of-bits, but as list-of-strings instead...

So, either we stick to the spec as it is, with a string-of-bits, of we go all the way to do a proper schema and use a boolean per bit.

**Comment by ymorin-orange** [comment]
> > > Is it better to keep bit strings such as:
> > > [bit-wise description]
> > > Or should we go instead for something more verbose like:
> > > [per-bit boolean property]
> > > Another alternative would be a list of enum:
> > > [enum]
> > The solution with a boolean property per bit is the best of the two; it is explicit and fits the spirit of a schema.
> > The enum proposal does looks kludgy and half-backed; it looks like the ETSI string-of-bits, but as list-of-strings instead...
> I agree that solution 2 (boolean property per bit) is the most elegant, but I'm really concerned about the message volume if we go this way,
> For example, with the `shared_with` field:
> 
> ```
> "shared_with": "0001100000"
> ```
> 
> ```
> "shared_with": {
>   "overlapping_lane_description_provided": false,
>   "multiple_lanes_treated_as_one_lane": false,
>   "other_non_motorized_traffic_types": false,
>   "individual_motorized_vehicle_traffic": true,
>   "bus_vehicle_traffic": true,
>   "taxi_vehicle_traffic": false,
>   "pedestrians_traffic": false,
>   "cyclist_vehicle_traffic": false,
>   "tracked_vehicle_traffic": false,
>   "pedestrian_traffic": false
> }
> ```
> 
> ```
> "shared_with": ["individualMotorizedVehicleTraffic", "busVehicleTraffic"]
> ```

OK, you sold me on it. Go for the list-of-strings, i.e. your proposal #3.

File: schema/dsrc/dsrc_schema_2-0-0.json [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+      "minItems": 1,
+      "maxItems": 9,
+      "items": {
+        "type": "object",
+        "required": [
+          "limit_type",
+          "speed"
+        ],
+        "properties": {
+          "type": {
```
Comment (ymorin-orange) [inline schema/dsrc/dsrc_schema_2-0-0.json:366]: You named the property `type`, but in the required list, it is `limit_type`. The name is DSRC is indeed `type`.

File: schema/dsrc/dsrc_schema_2-0-0.json [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+      "minimum": 0,
+      "maximum": 255
+    },
+    "lane_width": {
+      "type": "integer",
+      "description": "Conveys the width of a lane in LSB units of 1 cm. Maximum value for a lane is 327.67 meters in width.",
+      "minimum": 0,
+      "maximum": 32767
+    },
+    "ia5_string": {
```
Comment (ymorin-orange) [inline schema/dsrc/dsrc_schema_2-0-0.json:409]: I could not find a definition for IA5String in the DSRC... :-/

Comment (ymorin-orange) [reply schema/dsrc/dsrc_schema_2-0-0.json:409]: I could not find a definition for IA5String in the DSRC... :-/

Comment (ymorin-orange) [reply schema/dsrc/dsrc_schema_2-0-0.json:409]: As discussed IRL, drop the type and directly use a plain JSON string.

File: schema/mapem/mapem_schema_2-0-0.json [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+      "maximum": 1830297600000
+    },
+    "message": {
+      "type": "object",
+      "required": [
+        "protocol_version",
+        "station_id",
+        "msg_issue_revision"
+      ],
+      "properties": {
```
Comment (ymorin-orange) [inline schema/mapem/mapem_schema_2-0-0.json:60]: The [`regional`](https://forge.etsi.org/rep/ITS/asn1/is_ts103301/-/blob/v2.2.1/docs/ETSI-ITS-DSRC.md#mapdata) optional property is missing.

File: schema/mapem/mapem_schema_2-0-0.json [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+              }
+            }
+          }
+        },
+        "data_parameters": {
+          "type": "object",
+          "description": "Used to provide basic (static) information on how a map fragment was processed or determined.",
+          "properties": {
+            "process_method": {
+              "$ref": "../dsrc/dsrc_schema_2-0-0.json#/definitions/ia5_string"
```
Comment (ymorin-orange) [inline schema/mapem/mapem_schema_2-0-0.json:207]: As discussed in the DDSRC schema: make those plain JSON strings.

File: schema/mapem/mapem_schema_2-0-0.json [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+          "maxItems": 254,
+          "items": {
+            "type": "object",
+            "description": "Used to assign (or bind) a single RestrictionClassID data element to a list of all user classes to which it applies. A collection of these bindings is conveyed in the RestrictionClassList data frame in the MAP message to travelers. The established index is then used in the lane object of the MAP message, in the ConnectTo data frame, to qualify to whom a signal group ID applies when it is sent by the SPAT message about a movement.",
+            "required": [
+              "id",
+              "users"
+            ],
+            "properties": {
+              "property": {
```
Comment (ymorin-orange) [inline schema/mapem/mapem_schema_2-0-0.json:233]: Spurious `property` ?

File: schema/spatem/spatem_schema_2-0-0.json [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+                      "maxItems": 16,
+                      "items": {
+                        "type": "object",
+                        "description": "MovementEvent contains details about a single movement. It is used by the movement state to convey one of number of movements (typically occurring over a sequence of times) for a SignalGroupID.",
+                        "required": [
+                          "event_state"
+                        ],
+                        "properties": {
+                          "event_state": {
+                            "type": "integer",
```
Comment (ymorin-orange) [inline schema/spatem/spatem_schema_2-0-0.json:192]: The DSRC defines this as an integer, indeed, but the defintion is that of an enumerated type. In this case, maybe it makes more sense for us to use a string with en umerated values, like so:
```json
"event_state": {
  "type": string",
  "enum": [
    "unavailable",
    "dark",
    "stop-Then-Proceed",
    ...
  ]
}
```
Thoughts? If so, then do we want it in a separate commit, like we do for the bit-fields-turned-lists-of-string?

Comment (ymorin-orange) [reply schema/spatem/spatem_schema_2-0-0.json:192]: Hmm.. If we do so, there are so many other places where we'd need to change it too... Maybe that was not a great idea of me...

File: schema/mapem/mapem_schema_2-0-0.json [hunk 7]
```diff
@@ -243,7 +243,6 @@
                   "minItems": 1,
                   "maxItems": 16,
                   "items": {
-                    "$comment": "keep camelCase for ease of ASN.1 <> JSON translation, or use snake_case?",
```
Comment (ymorin-orange) [inline schema/mapem/mapem_schema_2-0-0.json]: You should keep that comment.

File: schema/dsrc/dsrc_schema_2-0-0.json [hunk 8]
```diff
@@ -249,12 +249,27 @@
...
+          "maneuverRightAllowed",
+          "maneuverUTurnAllowed",
+          "maneuverLeftTurnOnRedAllowed",
+          "maneuverRightTurnOnRedAllowed",
+          "maneuverLaneChangeAllowed",
+          "maneuverNoStoppingAllowed",
+          "yieldAllwaysRequired",
+          "goWithHalt",
+          "caution",
+          "reserved1"
```
Comment (ymorin-orange) [inline schema/dsrc/dsrc_schema_2-0-0.json:270]: As already discussed, drop the `reserved1`.

**Review by ymorin-orange — CHANGES_REQUESTED** [review]
A few minor omissions and tweaks; otherwise, titanic work you did there! ;-)

File: schema/dsrc/dsrc_schema_2-0-0.json (hunk 1 above)
Comment (mathieu1fb) [reply schema/dsrc/dsrc_schema_2-0-0.json:365]: Fixed.

File: schema/dsrc/dsrc_schema_2-0-0.json (hunk 2 above)
Comment (mathieu1fb) [reply schema/dsrc/dsrc_schema_2-0-0.json]: Dropped and replaced by a plain string.

File: schema/mapem/mapem_schema_2-0-0.json (hunk 4 above)
Comment (mathieu1fb) [reply schema/mapem/mapem_schema_2-0-0.json]: Done.

File: schema/mapem/mapem_schema_2-0-0.json (hunk 5 above)
Comment (mathieu1fb) [reply schema/mapem/mapem_schema_2-0-0.json]: My mistake indeed, it's been fixed!

File: schema/dsrc/dsrc_schema_2-0-0.json (hunk 8 above)
Comment (mathieu1fb) [reply schema/dsrc/dsrc_schema_2-0-0.json]: Removed.

File: schema/mapem/mapem_schema_2-0-0.json (hunk 7 above)
Comment (mathieu1fb) [reply schema/mapem/mapem_schema_2-0-0.json]: It was an oversight, I rebased so that it would not appear anymore in the original MAPEM commit.

File: schema/spatem/spatem_schema_2-0-0.json (hunk 6 above)
Comment (ymorin-orange) [reply schema/spatem/spatem_schema_2-0-0.json:192]: > maybe we should discuss the implications for our other JSON messages first and leave it for a next version?

Yes, keep it as-is for now.

**Review by ymorin-orange — APPROVED** [review]
